contracts/EthlanceJobs.sol

```diff
@@ -0,0 +1,843 @@
+      }
+    }
+
+    emit JobDrained(_jobId, _sender, _amounts);
+  }
+
+  /// @dev performAction(): Allows users to perform any generalized action
+  ///                       associated with a particular job, such as applying for it
+  /// @param _sender the sender of the transaction issuing the job (should be the same as msg.sender unless the txn is called by the meta tx relayer)
+  /// @param _jobId the index of the job
+  /// @param _ipfsHash the IPFS hash corresponding to a JSON object which contains the details of the action being performed (see docs for schema details)
+  function performAction(
+                         address _sender,
+                         uint _jobId,
+                         string memory _ipfsHash)
+    public
+    senderIsValid(_sender)
+    validateJobArrayIndex(_jobId)
+  {
+    emit ActionPerformed(_jobId, _sender, _ipfsHash); // The _ipfsHash string is emitted in an event for easy off-chain consumption
+  }
+
+  /// @dev fulfillJob(): Allows users to fulfill the job to get paid out
+  /// @param _sender the sender of the transaction issuing the job (should be the same as msg.sender unless the txn is called by the meta tx relayer)
+  /// @param _jobId the index of the job
+  /// @param _fulfillers the array of addresses which will receive payouts for the submission
+  /// @param _ipfsHash the IPFS hash corresponding to a JSON object which contains the details of the submission (see docs for schema details)
+  function fulfillJob(
+                      address _sender,
+                      uint _jobId,
+                      address payable[] memory  _fulfillers,
+                      string memory _ipfsHash,
+                      uint _amount)
+    public
+    senderIsValid(_sender)
+    validateJobArrayIndex(_jobId)
+  {
+    require(contains(jobs[_jobId].hiredCandidates, _sender));
+
+    require(_fulfillers.length > 0); // Submissions with no fulfillers would mean no one gets paid out
+
+    jobs[_jobId].fulfillments.push(Fulfillment(_fulfillers, _sender, _amount));
+
+    emit JobFulfilled(_jobId,
+                      (jobs[_jobId].fulfillments.length - 1),
+                      _fulfillers,
+                      _ipfsHash, // The _ipfsHash string is emitted in an event for easy off-chain consumption
+                      _sender,
+                      _amount);
+  }
+
+  /// @dev updateFulfillment(): Allows the submitter of a fulfillment to update their submission
+  /// @param _sender the sender of the transaction issuing the job (should be the same as msg.sender unless the txn is called by the meta tx relayer)
+  /// @param _jobId the index of the job
+  /// @param _fulfillmentId the index of the fulfillment
+  /// @param _fulfillers the new array of addresses which will receive payouts for the submission
+  /// @param _ipfsHash the new IPFS hash corresponding to a JSON object which contains the details of the submission (see docs for schema details)
+  function updateFulfillment(
```

Comment:
not needed for Jobs, and I think better would be if we have `cancelFulfillment` function, to be able to cancel the invoice in case of the mistake.

Reply:
what is the procedure then for modifying the files you are sending to the employer? It will end up with the job pointing to the first file set you sent.

Reply:
Fulfillment is just invoice, user can send as many invoices & messages to the employer as he needs. No need to be able to edit previous messages. It just creates complexity without much benefit.